crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs

```diff
@@ -713,73 +642,74 @@ impl Space {
             space,
             from_agent,
             basis,
-            //remote_agent_count,
+            remote_agent_count,
             timeout_ms,
             //as_race,
             //race_timeout_ms,
             payload,
             ..
         } = input;
+        let remote_agent_count = remote_agent_count.unwrap();
         let timeout_ms = timeout_ms.unwrap();
+        let stage_1_timeout_ms = timeout_ms / 2;
 
-        // TODO - we cannot write proper logic here until we have a
-        //        proper peer discovery mechanism. Instead, let's
-        //        give it 100 ms max to see if there is any agent
-        //        other than us - prefer that, or fall back to
-        //        just reflecting the msg to ourselves.
-
-        let i_s = self.i_s.clone();
-        Ok(async move {
-            let mut to_agent = from_agent.clone();
-            'search_loop: for _ in 0..5 {
-                if let Ok(agent_list) = i_s
-                    .list_online_agents_for_basis_hash(
+        // as an optimization - request to all local joins
+        // but don't count that toward our request total
+        let local_all = self
+            .local_joined_agents
+            .iter()
+            .map(|agent| {
+                let agent = agent.clone();
+                self.evt_sender
+                    .call(
                         space.clone(),
+                        agent.clone(),
                         from_agent.clone(),
-                        basis.clone(),
+                        payload.clone(),
                     )
-                    .await
-                {
-                    for a in agent_list {
-                        if a != from_agent {
-                            to_agent = a;
-                            break 'search_loop;
-                        }
-                    }
-                }
+                    .then(|r| async move { (r, agent) })
+            })
+            .collect::<Vec<_>>();
 
-                tokio::time::delay_for(std::time::Duration::from_millis(20)).await;
-            }
+        let remote_fut = discover::message_neighborhood(
+            self,
+            from_agent.clone(),
+            remote_agent_count,
+            stage_1_timeout_ms,
+            timeout_ms,
+            basis,
+            wire::Wire::call(
+                space.clone(),
+                from_agent.clone(),
+                from_agent,
+                payload.into(),
+            ),
+            |a, w| match w {
+                wire::Wire::CallResp(c) => Ok(actor::RpcMultiResponse {
+                    agent: a,
+                    response: c.data.into(),
+                }),
+                _ => Err(()),
+            },
+        );
 
-            let mut out = Vec::new();
+        Ok(async move {
+            let mut out: Vec<actor::RpcMultiResponse> = futures::future::join_all(local_all)
```

Comment:
Probably not an issue now but just noting this for the future. `join_all()` will cancel all futures if any fail. Not sure if that's desired behaviour but probably fine for now.

Reply:
Fail in terms of any futures that panic!, though... which, if our code is panicking, we have bigger problems.